### PR TITLE
views: validation error catcher

### PR DESCRIPTION
**Pull request by hachreak**
Signed-off-by: Leonardo Rossi <leonardo.r@cern.ch>

### CODE REVIEW

File: invenio_records_rest/utils.py [hunk 1]
```diff
@@ -34,6 +34,18 @@
 from invenio_records.api import Record
 from werkzeug.routing import BaseConverter, BuildError
 from werkzeug.utils import cached_property, import_string
+from functools import wraps
+from jsonschema.exceptions import ValidationError
+
+
+def catch_record_validation_error(func):
```
Comment (lnielsen) [inline invenio_records_rest/utils.py]: Probably better suited as an error handler on the blueprint: http://flask.pocoo.org/docs/0.10/api/#flask.Blueprint.errorhandler

File: invenio_records_rest/views.py [hunk 2]
```diff
@@ -69,6 +67,12 @@ def create_blueprint(endpoints):
         for rule in create_url_rules(endpoint, **options):
             blueprint.add_url_rule(**rule)
 
+    # catch record validation errors
+    @blueprint.errorhandler(ValidationError)
+    def validation_error(error):
+        """Catch validation errors."""
+        return "Validation error", 400
```
Comment (jirikuncar) [inline invenio_records_rest/views.py]: We should be careful with response body. We should probably return body of a  `RESTException`.

Comment (hachreak) [inline invenio_records_rest/views.py]: do you mean like this?
```python
return RESTException(description="Validation error").get_body(), 400
```

Comment (jirikuncar) [inline invenio_records_rest/views.py]: IMHO just `RESTException(description="Validation error")` see https://github.com/pallets/werkzeug/blob/master/werkzeug/exceptions.py#L75-L162

Comment (lnielsen) [inline invenio_records_rest/views.py]: What about a generic ``RESTValidatinoError`` in that models responses according to  https://zenodo.org/dev#restapi-errors. I.e.:

```json
{
    "message": "Validation error",
    "status": 400,
    "errors": [
        {"code": 10, "message":"Not a valid choice", "field": "access_right"}
    ]
}
```

See https://github.com/inveniosoftware/invenio-rest/pull/34/files

```python
from invenio_rest.errors import RESTValidationError, FieldError

raise RESTValidationError(errors=[FieldError('access_right', 'Not a valid choice', code=10)])
```

File: invenio_records_rest/views.py [hunk 3]
```diff
@@ -69,6 +68,12 @@ def create_blueprint(endpoints):
         for rule in create_url_rules(endpoint, **options):
             blueprint.add_url_rule(**rule)
 
+    # catch record validation errors
+    @blueprint.errorhandler(ValidationError)
+    def validation_error(error):
+        """Catch validation errors."""
+        return RESTException(description="Validation error"), 400
```
Comment (lnielsen) [inline invenio_records_rest/views.py]: What about a generic ``RESTValidatinoError`` in that models responses according to  https://zenodo.org/dev#restapi-errors. I.e.:

```json
{
    "message": "Validation error",
    "status": 400,
    "errors": [
        {"code": 10, "message":"Not a valid choice", "field": "access_right"}
    ]
}
```

See https://github.com/inveniosoftware/invenio-rest/pull/34/files

```python
from invenio_rest.errors import RESTValidationError, FieldError

raise RESTValidationError(errors=[FieldError('access_right', 'Not a valid choice', code=10)])
```

The RESTValidationError will be caught by Invenio-REST error handlers.

File: invenio_records_rest/views.py [hunk 4]
```diff
@@ -69,6 +68,12 @@ def create_blueprint(endpoints):
         for rule in create_url_rules(endpoint, **options):
             blueprint.add_url_rule(**rule)
 
+    # catch record validation errors
+    @blueprint.errorhandler(ValidationError)
+    def validation_error(error):
+        """Catch validation errors."""
+        return RESTValidationError().get_response()
```
Comment (jirikuncar) [inline invenio_records_rest/views.py]: `raise RESTValidationError()`?

Comment (hachreak) [inline invenio_records_rest/views.py]: I tried to raise but it returns error: ` TypeError: exceptions must be old-style classes or derived from BaseException, not Response`

**Comment by lnielsen** [comment]
LGTM - please rebase.

**Pull request merged**